url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/rtree/SecurityTree.java

```diff
@@ -0,0 +1,164 @@
+package spatialindex.rtree;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.math.BigInteger;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Set;
+
+public class SecurityTree {
+	public String filePath = "";
+	private RTree rtree;
+	SecurityNode root;
+	private HashMap<Integer, SecurityNode> SecurityNodeMap = new HashMap<Integer, SecurityNode>();
+
+	/**
+	 * if this contains a value means it should be load from the file, save (ID,
+	 * pos in index file) pairs
+	 */
+	private HashMap<Integer, Long> SecurityNodeMapPos = new HashMap<Integer, Long>();
+	private HashMap<Integer, Integer> SecurityNodeMapLength = new HashMap<Integer, Integer>();
+
+	int level;
+
+	public SecurityTree(RTree _rtree, String _filePath, boolean isLoad) throws IndexOutOfBoundsException, Exception {
+		System.out.println("building security info");
+		long startTime = System.currentTimeMillis();
+		filePath = _filePath;
+		if(!isLoad){
+			rtree = _rtree;
+			buildSecurityTree(this, _rtree);
+		}else{
+			rtree = _rtree;
+			readFromFile(rtree, filePath);
+			root = getSecurityNode(rtree.m_rootID);
+			//System.out.println(new BigInteger(root.getEntaValue().getBytes("ISO-8859-1")));
+		}
+		long endTime = System.currentTimeMillis();
+		long cpuTime = endTime - startTime;
+		System.out.println("build security info finished, time used:" + cpuTime + "ms");
+	}
+
+	public void buildSecurityTree(SecurityTree srtree, RTree _rtree) throws IndexOutOfBoundsException, Exception {
+		root = SecurityNode.buildSecurityNode(srtree, rtree);
+		level = root.getLevel();
+	}
+
+	public void saveSecurityNode(SecurityNode n, int id) {
+//		if (n.getLevel() > 1) {
+//			System.out.println("sNode[" + id + "] saved, lv=" + n.getLevel());
+//		}
+//		if(n.getLevel() > 1) 
+//			SecurityNodeMap.put(new Integer(id), n);
+	}
+
+	public SecurityNode getSecurityNode(int id) {
+		SecurityNode snode = SecurityNodeMap.get(new Integer(id));
+		if (snode == null) {
+			Long pos = SecurityNodeMapPos.get(new Integer(id));
+			if (pos == null) {
+				return null;
+			} else {
+				int length = SecurityNodeMapLength.get(new Integer(id)).intValue();
+				//load from file
+				try {
+					RandomAccessFile rfile = new RandomAccessFile(new File(filePath + ".sdat"), "r");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *rfile*. The resource is referenced at lines 73, 75. The resource is closed at line 76. There are other execution paths that do not contain closure statements, for example, when RandomAccessFile.seek throws an exception. Ensure *rfile* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).